File: src/34. Find First and Last Position of Element in Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

export function searchRange(nums, target) {
  return [searchFirst(nums, target), searchLast(nums, target)]
}

function searchFirst(nums, target) {
  let left = 0,
    right = nums.length
  while (left < right) {
    let mid = Math.floor((left + right) / 2)
    if (nums[mid] >= target) right = mid
    else left = mid + 1
  }
  if (left === nums.length || nums[left] !== target) return -1
  return left
}

function searchLast(nums, target) {
  let left = 0,
    right = nums.length
  while (left < right) {
    let mid = Math.floor((left + right) / 2)
    if (nums[mid] > target) right = mid
    else left = mid + 1
  }
  left--
  if (left < 0 || nums[left] !== target) return -1
  return left
}

File: src/34. Find First and Last Position of Element in Sorted Array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def searchFirst() -> int:
            left, right = 0, len(nums)
            while left < right:
                mid = (left + right) // 2
                if nums[mid] >= target:
                    right = mid
                else:
                    left = mid + 1
            return -1 if left >= len(nums) or nums[left] != target else left

        def searchLast() -> int:
            left, right = 0, len(nums)
            while left < right:
                mid = (left + right) // 2
                if nums[mid] <= target:
                    left = mid + 1
                else:
                    right = mid
            left -= 1
            return -1 if left < 0 or nums[left] != target else left

        return [searchFirst(), searchLast()]